File: api/src/smart-capture/smart-capture.service.ts
import { Injectable, Logger, BadRequestException } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { firstValueFrom } from 'rxjs';
import { PrismaService } from '../database/prisma.service';
import { LeadsService } from '../leads/leads.service';
import { ContactsService } from '../contacts/contacts.service';
import { AccountsService } from '../accounts/accounts.service';
import { NotesService } from '../notes/notes.service';
import { AnthropicService } from '../anthropic/anthropic.service';
import {
  CaptureMode,
  CrmEntityType,
  ExtractedContactDto,
  ExtractedTextDto,
  ExtractedReceiptDto,
  SmartCaptureResultDto,
  CreateFromCaptureDto,
  SmartCaptureConfigDto,
} from './dto';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const FormData = require('form-data');

interface UploadedFile {
  buffer: Buffer;
  originalname: string;
  mimetype: string;
  size: number;
}

// Default smart capture configuration
const DEFAULT_CONFIG: SmartCaptureConfigDto = {
  enableBusinessCardScan: true,
  enableDocumentScan: true,
  enableHandwritingRecognition: true,
  enableReceiptScan: true,
  autoCreateLeadFromBusinessCard: false,
  aiEnhancedExtraction: true,
  defaultNoteEntityType: 'lead',
};

@Injectable()
export class SmartCaptureService {
  private readonly logger = new Logger(SmartCaptureService.name);
  private readonly pageIndexUrl: string;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
    private readonly prisma: PrismaService,
    private readonly leadsService: LeadsService,
    private readonly contactsService: ContactsService,
    private readonly accountsService: AccountsService,
    private readonly notesService: NotesService,
    private readonly anthropicService: AnthropicService,
  ) {
    this.pageIndexUrl = this.configService.get<string>('PAGEINDEX_SERVICE_URL') || 'http://localhost:8001';
    this.logger.log(`Smart Capture service initialized, PageIndex URL: ${this.pageIndexUrl}`);
  }

  /**
   * Get smart capture configuration from system config
   */
  async getConfig(): Promise<SmartCaptureConfigDto> {
    try {
      const settings = await this.prisma.systemConfig.findUnique({
        where: { key: 'smart_capture_config' },
      });
      if (settings && settings.value) {
        return { ...DEFAULT_CONFIG, ...JSON.parse(settings.value) };
      }
    } catch (error) {
      this.logger.warn('Could not load smart capture config, using defaults');
    }
    return DEFAULT_CONFIG;
  }

  /**
   * Save smart capture configuration
   */
  async saveConfig(config: SmartCaptureConfigDto): Promise<SmartCaptureConfigDto> {
    const existing = await this.prisma.systemConfig.findUnique({
      where: { key: 'smart_capture_config' },
    });

    if (existing) {
      await this.prisma.systemConfig.update({
        where: { key: 'smart_capture_config' },
        data: { value: JSON.stringify(config) },
      });
    } else {
      await this.prisma.systemConfig.create({
        data: {
          key: 'smart_capture_config',
          value: JSON.stringify(config),
          category: 'smart_capture',
          type: 'json',
          label: 'Smart Capture Configuration',
          description: 'Configuration for Smart Camera & Notepad features',
        },
      });
    }

    return this.getConfig();
  }

  /**
   * Process a business card image
   */
  async processBusinessCard(
    file: UploadedFile,
    userId: string,
    organizationId: string,
    autoCreate?: boolean,
  ): Promise<SmartCaptureResultDto> {
    this.logger.log(`Processing business card: ${file.originalname}`);

    try {
      const formData = new FormData();
      formData.append('file', file.buffer, {
        filename: file.originalname,
        contentType: file.mimetype,
      });

      const response = await firstValueFrom(
        this.httpService.post(`${this.pageIndexUrl}/ocr/business-card`, formData, {
          headers: formData.getHeaders(),
          timeout: 60000,
        }),
      );

      const ocrResult = response.data;
      const extractedData: ExtractedContactDto = ocrResult.contact;

      // Optionally enhance with AI
      const config = await this.getConfig();
      let aiEnhanced = false;

      if (config.aiEnhancedExtraction && extractedData) {
        try {
          const enhanced = await this.enhanceContactWithAI(extractedData);
          Object.assign(extractedData, enhanced);
          aiEnhanced = true;
        } catch (error) {
          this.logger.warn('AI enhancement failed, using raw OCR data');
        }
      }

      const result: SmartCaptureResultDto = {
        success: ocrResult.success,
        mode: CaptureMode.BUSINESS_CARD,
        extractedData,
        aiEnhanced,
      };

      // Auto-create lead/contact if requested
      if (autoCreate || config.autoCreateLeadFromBusinessCard) {
        try {
          const lead = await this.createLeadFromContact(extractedData, userId, organizationId);
          result.createdEntity = {
            type: CrmEntityType.LEAD,
            id: lead.id,
            name: `${lead.firstName || ''} ${lead.lastName || ''}`.trim(),
          };
        } catch (error) {
          this.logger.error('Auto-create lead failed:', error);
        }
      }

      return result;
    } catch (error) {
      this.logger.error(`Business card processing failed: ${error.message}`);
      return {
        success: false,
        mode: CaptureMode.BUSINESS_CARD,
        extractedData: {},
        error: error.message,
      };
    }
  }

  /**
   * Process a document image with OCR
   */
  async processDocument(file: UploadedFile, userId: string): Promise<SmartCaptureResultDto> {
    this.logger.log(`Processing document: ${file.originalname}`);

    try {
      const formData = new FormData();
      formData.append('file', file.buffer, {
        filename: file.originalname,
        contentType: file.mimetype,
      });

      const response = await firstValueFrom(
        this.httpService.post(`${this.pageIndexUrl}/ocr/image`, formData, {
          headers: formData.getHeaders(),
          timeout: 60000,
        }),
      );

      const ocrResult = response.data;
      const extractedData: ExtractedTextDto = {
        text: ocrResult.text,
        pageCount: ocrResult.pageCount,
        hasHandwriting: ocrResult.hasHandwriting,
        confidence: ocrResult.confidence,
        wordCount: ocrResult.wordCount,
      };

      return {
        success: true,
        mode: CaptureMode.DOCUMENT,
        extractedData,
      };
    } catch (error) {
      this.logger.error(`Document processing failed: ${error.message}`);
      return {
        success: false,
        mode: CaptureMode.DOCUMENT,
        extractedData: { text: '' },
        error: error.message,
      };
    }
  }

  /**
   * Process handwritten notes
   */
  async processHandwritten(
    file: UploadedFile,
    userId: string,
    organizationId: string,
    options?: {
      createNote?: boolean;
      linkedEntityId?: string;
      linkedEntityType?: string;
      noteTitle?: string;
    },
  ): Promise<SmartCaptureResultDto> {
    this.logger.log(`Processing handwritten notes: ${file.originalname}`);

    try {
      const formData = new FormData();
      formData.append('file', file.buffer, {
        filename: file.originalname,
        contentType: file.mimetype,
      });

      const response = await firstValueFrom(
        this.httpService.post(`${this.pageIndexUrl}/ocr/handwritten`, formData, {
          headers: formData.getHeaders(),
          timeout: 60000,
        }),
      );

      const ocrResult = response.data;
      const extractedData: ExtractedTextDto = {
        text: ocrResult.text,
        hasHandwriting: ocrResult.isMainlyHandwritten,
        confidence: ocrResult.confidence,
        wordCount: ocrResult.wordCount,
      };

      const result: SmartCaptureResultDto = {
        success: true,
        mode: CaptureMode.HANDWRITTEN,
        extractedData,
      };

      // Create note if requested
      if (options?.createNote && extractedData.text) {
        try {
          const noteData: any = {
            title: options.noteTitle || `Handwritten Note - ${new Date().toLocaleDateString()}`,
            body: extractedData.text,
          };

          // Link to entity if specified
          if (options.linkedEntityId && options.linkedEntityType) {
            switch (options.linkedEntityType.toLowerCase()) {
              case 'lead':
                noteData.leadId = options.linkedEntityId;
                break;
              case 'contact':
                noteData.contactId = options.linkedEntityId;
                break;
              case 'account':
                noteData.accountId = options.linkedEntityId;
                break;
              case 'opportunity':
                noteData.opportunityId = options.linkedEntityId;
                break;
            }
          }

          const note = await this.notesService.createNote(noteData, userId, organizationId);
          result.createdEntity = {
            type: CrmEntityType.NOTE,
            id: note.id,
            name: note.title || 'Note',
          };
        } catch (error) {
          this.logger.error('Create note failed:', error);
        }
      }

      return result;
    } catch (error) {
      this.logger.error(`Handwritten processing failed: ${error.message}`);
      return {
        success: false,
        mode: CaptureMode.HANDWRITTEN,
        extractedData: { text: '' },
        error: error.message,
      };
    }
  }

  /**
   * Process receipt image
   */
  async processReceipt(file: UploadedFile, userId: string): Promise<SmartCaptureResultDto> {
    this.logger.log(`Processing receipt: ${file.originalname}`);

    try {
      const formData = new FormData();
      formData.append('file', file.buffer, {
        filename: file.originalname,
        contentType: file.mimetype,
      });

      const response = await firstValueFrom(
        this.httpService.post(`${this.pageIndexUrl}/ocr/receipt`, formData, {
          headers: formData.getHeaders(),
          timeout: 60000,
        }),
      );

      const ocrResult = response.data;
      const extractedData: ExtractedReceiptDto = ocrResult.receipt;

      return {
        success: ocrResult.success,
        mode: CaptureMode.RECEIPT,
        extractedData,
      };
    } catch (error) {
      this.logger.error(`Receipt processing failed: ${error.message}`);
      return {
        success: false,
        mode: CaptureMode.RECEIPT,
        extractedData: {},
        error: error.message,
      };
    }
  }

  /**
   * Create CRM entity from extracted data
   */
  async createEntityFromCapture(
    dto: CreateFromCaptureDto,
    userId: string,
    organizationId: string,
  ): Promise<SmartCaptureResultDto> {
    this.logger.log(`Creating ${dto.entityType} from captured data`);

    try {
      let createdEntity: any;
      const data = { ...dto.extractedData, ...dto.additionalFields };

      switch (dto.entityType) {
        case CrmEntityType.LEAD:
          createdEntity = await this.createLeadFromContact(data as ExtractedContactDto, userId, organizationId);
          break;

        case CrmEntityType.CONTACT:
          createdEntity = await this.createContactFromExtracted(data as ExtractedContactDto, userId, organizationId);
          break;

        case CrmEntityType.ACCOUNT:
          createdEntity = await this.createAccountFromExtracted(data, userId, organizationId);
          break;

        case CrmEntityType.NOTE:
          createdEntity = await this.notesService.createNote(
            {
              title: data.title || 'Captured Note',
              body: data.text || data.body || '',
              leadId: dto.linkedEntityType === 'lead' ? dto.linkedEntityId : undefined,
              contactId: dto.linkedEntityType === 'contact' ? dto.linkedEntityId : undefined,
              accountId: dto.linkedEntityType === 'account' ? dto.linkedEntityId : undefined,
              opportunityId: dto.linkedEntityType === 'opportunity' ? dto.linkedEntityId : undefined,
            },
            userId,
            organizationId,
          );
          break;

        default:
          throw new BadRequestException(`Unsupported entity type: ${dto.entityType}`);
      }

      return {
        success: true,
        mode: CaptureMode.DOCUMENT,
        extractedData: data,
        createdEntity: {
          type: dto.entityType,
          id: createdEntity.id,
          name: this.getEntityName(createdEntity, dto.entityType),
        },
      };
    } catch (error) {
      this.logger.error(`Create entity failed: ${error.message}`);
      return {
        success: false,
        mode: CaptureMode.DOCUMENT,
        extractedData: dto.extractedData,
        error: error.message,
      };
    }
  }

  /**
   * Enhance extracted contact data with AI
   */
  private async enhanceContactWithAI(contact: ExtractedContactDto): Promise<Partial<ExtractedContactDto>> {
    const prompt = `Clean up and enhance this extracted business card data. Fix any OCR errors, normalize phone numbers, and fill in missing fields if you can reasonably infer them.

Input data:
${JSON.stringify(contact, null, 2)}

Return a JSON object with the corrected/enhanced fields only. Do not add fields that cannot be reasonably inferred.`;

    try {
      const response = await this.anthropicService.generateChatCompletion({
        messages: [{ role: 'user', content: prompt }],
      });

      // Parse JSON from response
      const jsonMatch = response.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        return JSON.parse(jsonMatch[0]);
      }
    } catch (error) {
      this.logger.warn('AI enhancement failed:', error.message);
    }

    return {};
  }

  /**
   * Create a lead from extracted contact data
   */
  private async createLeadFromContact(contact: ExtractedContactDto, userId: string, organizationId: string) {
    return this.leadsService.create(
      userId,
      {
        firstName: contact.firstName || 'Unknown',
        lastName: contact.lastName || '',
        email: contact.email,
        phone: contact.phone || contact.mobilePhone,
        company: contact.company,
        title: contact.jobTitle,
        website: contact.website,
        leadSource: 'OTHER',
        description: `Created from business card scan. Address: ${contact.address || 'N/A'}${contact.mobilePhone ? ` Mobile: ${contact.mobilePhone}` : ''}`,
      },
      organizationId,
    );
  }

  /**
   * Create a contact from extracted data
   */
  private async createContactFromExtracted(contact: ExtractedContactDto, userId: string, organizationId: string) {
    // First try to find or create an account if company is present
    let accountId: string | undefined;

    if (contact.company) {
      try {
        const existingAccount = await this.prisma.account.findFirst({
          where: {
            ownerId: userId,
            organizationId,
            name: { contains: contact.company, mode: 'insensitive' },
          },
        });

        if (existingAccount) {
          accountId = existingAccount.id;
        } else {
          const newAccount = await this.accountsService.createAccount(
            {
              name: contact.company,
              type: 'PROSPECT',
              website: contact.website,
              phone: contact.phone,
            },
            userId,
            organizationId,
          );
          accountId = newAccount.id;
        }
      } catch (error) {
        this.logger.warn('Could not create/find account for contact:', error.message);
      }
    }

    // Build contact data - accountId is required so throw if not available
    if (!accountId) {
      throw new Error('Cannot create contact without an associated account');
    }

    return this.contactsService.createContact(
      {
        firstName: contact.firstName || 'Unknown',
        lastName: contact.lastName || '',
        email: contact.email,
        phone: contact.phone,
        mobilePhone: contact.mobilePhone,
        title: contact.jobTitle,
        department: contact.department,
        accountId,
        description: `Created from business card scan.`,
      },
      userId,
      organizationId,
    );
  }

  /**
   * Create an account from extracted data
   */
  private async createAccountFromExtracted(data: any, userId: string, organizationId: string) {
    return this.accountsService.createAccount(
      {
        name: data.company || data.merchantName || 'Unknown Company',
        type: data.type || 'PROSPECT',
        website: data.website,
        phone: data.phone || data.merchantPhone,
        billingStreet: data.address || data.merchantAddress,
        description: data.description,
      },
      userId,
      organizationId,
    );
  }

  /**
   * Get display name for an entity
   */
  private getEntityName(entity: any, type: CrmEntityType): string {
    switch (type) {
      case CrmEntityType.LEAD:
      case CrmEntityType.CONTACT:
        return `${entity.firstName || ''} ${entity.lastName || ''}`.trim() || 'Unnamed';
      case CrmEntityType.ACCOUNT:
        return entity.name || 'Unnamed Account';
      case CrmEntityType.NOTE:
        return entity.title || 'Note';
      default:
        return 'Unknown';
    }
  }

  /**
   * Check OCR service availability
   */
  async checkOcrAvailability(): Promise<{
    available: boolean;
    capabilities: string[];
  }> {
    try {
      const response = await firstValueFrom(
        this.httpService.get(`${this.pageIndexUrl}/capabilities`, { timeout: 5000 }),
      );

      const caps = response.data.capabilities;
      const capabilities: string[] = [];

      if (caps.ocr) capabilities.push('ocr');
      if (caps.business_card_scan) capabilities.push('business_card');
      if (caps.handwriting_recognition) capabilities.push('handwriting');
      if (caps.receipt_scan) capabilities.push('receipt');
      if (caps.table_extraction) capabilities.push('table_extraction');

      return {
        available: caps.ocr === true,
        capabilities,
      };
    } catch (error) {
      this.logger.error('OCR service check failed:', error.message);
      return {
        available: false,
        capabilities: [],
      };
    }
  }
}

File: salesos_mobile/lib/core/services/smart_capture_service.dart
import 'dart:io';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:image_picker/image_picker.dart';
import '../network/api_client.dart';

/// Capture modes for Smart Capture
enum CaptureMode {
  businessCard,
  document,
  handwritten,
  receipt,
}

/// CRM entity types for creating entities from captured data
enum CrmEntityType {
  lead,
  contact,
  account,
  note,
}

/// Result of a smart capture operation
class SmartCaptureResult {
  final bool success;
  final CaptureMode mode;
  final Map<String, dynamic> extractedData;
  final bool aiEnhanced;
  final CreatedEntity? createdEntity;
  final String? error;

  SmartCaptureResult({
    required this.success,
    required this.mode,
    required this.extractedData,
    this.aiEnhanced = false,
    this.createdEntity,
    this.error,
  });

  factory SmartCaptureResult.fromJson(Map<String, dynamic> json) {
    return SmartCaptureResult(
      success: json['success'] ?? false,
      mode: _parseCaptureMode(json['mode']),
      extractedData: json['extractedData'] ?? {},
      aiEnhanced: json['aiEnhanced'] ?? false,
      createdEntity: json['createdEntity'] != null
          ? CreatedEntity.fromJson(json['createdEntity'])
          : null,
      error: json['error'],
    );
  }

  static CaptureMode _parseCaptureMode(String? mode) {
    switch (mode) {
      case 'business_card':
        return CaptureMode.businessCard;
      case 'document':
        return CaptureMode.document;
      case 'handwritten':
        return CaptureMode.handwritten;
      case 'receipt':
        return CaptureMode.receipt;
      default:
        return CaptureMode.document;
    }
  }
}

/// Entity created from captured data
class CreatedEntity {
  final CrmEntityType type;
  final String id;
  final String? name;

  CreatedEntity({
    required this.type,
    required this.id,
    this.name,
  });

  factory CreatedEntity.fromJson(Map<String, dynamic> json) {
    return CreatedEntity(
      type: _parseEntityType(json['type']),
      id: json['id'] ?? '',
      name: json['name'],
    );
  }

  static CrmEntityType _parseEntityType(String? type) {
    switch (type) {
      case 'lead':
        return CrmEntityType.lead;
      case 'contact':
        return CrmEntityType.contact;
      case 'account':
        return CrmEntityType.account;
      case 'note':
        return CrmEntityType.note;
      default:
        return CrmEntityType.lead;
    }
  }
}

/// Extracted contact information from business card
class ExtractedContact {
  final String? firstName;
  final String? lastName;
  final String? fullName;
  final String? jobTitle;
  final String? company;
  final String? department;
  final String? email;
  final String? phone;
  final String? mobilePhone;
  final String? fax;
  final String? website;
  final String? address;
  final double? confidence;

  ExtractedContact({
    this.firstName,
    this.lastName,
    this.fullName,
    this.jobTitle,
    this.company,
    this.department,
    this.email,
    this.phone,
    this.mobilePhone,
    this.fax,
    this.website,
    this.address,
    this.confidence,
  });

  factory ExtractedContact.fromJson(Map<String, dynamic> json) {
    return ExtractedContact(
      firstName: json['firstName'],
      lastName: json['lastName'],
      fullName: json['fullName'],
      jobTitle: json['jobTitle'],
      company: json['company'],
      department: json['department'],
      email: json['email'],
      phone: json['phone'],
      mobilePhone: json['mobilePhone'],
      fax: json['fax'],
      website: json['website'],
      address: json['address'],
      confidence: json['confidence']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {
        'firstName': firstName,
        'lastName': lastName,
        'fullName': fullName,
        'jobTitle': jobTitle,
        'company': company,
        'department': department,
        'email': email,
        'phone': phone,
        'mobilePhone': mobilePhone,
        'fax': fax,
        'website': website,
        'address': address,
        'confidence': confidence,
      };

  bool get hasData =>
      firstName != null ||
      lastName != null ||
      email != null ||
      phone != null ||
      company != null;
}

/// OCR service capabilities
class OcrCapabilities {
  final bool available;
  final List<String> capabilities;

  OcrCapabilities({
    required this.available,
    required this.capabilities,
  });

  factory OcrCapabilities.fromJson(Map<String, dynamic> json) {
    return OcrCapabilities(
      available: json['available'] ?? false,
      capabilities: List<String>.from(json['capabilities'] ?? []),
    );
  }

  bool get hasBusinessCard => capabilities.contains('business_card');
  bool get hasHandwriting => capabilities.contains('handwriting');
  bool get hasReceipt => capabilities.contains('receipt');
}

/// Smart Capture Service for OCR and AI-powered data extraction
class SmartCaptureService {
  final ApiClient _apiClient;
  final ImagePicker _imagePicker = ImagePicker();

  SmartCaptureService(this._apiClient);

  /// Check OCR service availability and capabilities
  Future<OcrCapabilities> getCapabilities() async {
    try {
      final response = await _apiClient.get('/smart-capture/capabilities');
      return OcrCapabilities.fromJson(response.data);
    } catch (e) {
      return OcrCapabilities(available: false, capabilities: []);
    }
  }

  /// Capture image from camera
  Future<File?> captureFromCamera() async {
    try {
      final XFile? image = await _imagePicker.pickImage(
        source: ImageSource.camera,
        maxWidth: 2048,
        maxHeight: 2048,
        imageQuality: 90,
      );
      if (image != null) {
        return File(image.path);
      }
    } catch (e) {
      // Silently ignore
    }
    return null;
  }

  /// Pick image from gallery
  Future<File?> pickFromGallery() async {
    try {
      final XFile? image = await _imagePicker.pickImage(
        source: ImageSource.gallery,
        maxWidth: 2048,
        maxHeight: 2048,
        imageQuality: 90,
      );
      if (image != null) {
        return File(image.path);
      }
    } catch (e) {
      // Silently ignore
    }
    return null;
  }

  /// Process business card image
  Future<SmartCaptureResult> processBusinessCard(
    File imageFile, {
    bool autoCreate = false,
  }) async {
    try {
      final response = await _apiClient.uploadFile(
        '/smart-capture/business-card?autoCreate=$autoCreate',
        filePath: imageFile.path,
        fieldName: 'file',
      );
      return SmartCaptureResult.fromJson(response.data);
    } catch (e) {
      return SmartCaptureResult(
        success: false,
        mode: CaptureMode.businessCard,
        extractedData: {},
        error: e.toString(),
      );
    }
  }

  /// Process document image with OCR
  Future<SmartCaptureResult> processDocument(File imageFile) async {
    try {
      final response = await _apiClient.uploadFile(
        '/smart-capture/document',
        filePath: imageFile.path,
        fieldName: 'file',
      );
      return SmartCaptureResult.fromJson(response.data);
    } catch (e) {
      return SmartCaptureResult(
        success: false,
        mode: CaptureMode.document,
        extractedData: {},
        error: e.toString(),
      );
    }
  }

  /// Process handwritten notes
  Future<SmartCaptureResult> processHandwritten(
    File imageFile, {
    bool createNote = false,
    String? linkedEntityId,
    String? linkedEntityType,
    String? noteTitle,
  }) async {
    try {
      final response = await _apiClient.uploadFile(
        '/smart-capture/handwritten',
        filePath: imageFile.path,
        fieldName: 'file',
        data: {
          'createNote': createNote,
          'linkedEntityId': ?linkedEntityId,
          'linkedEntityType': ?linkedEntityType,
          'noteTitle': ?noteTitle,
        },
      );
      return SmartCaptureResult.fromJson(response.data);
    } catch (e) {
      return SmartCaptureResult(
        success: false,
        mode: CaptureMode.handwritten,
        extractedData: {},
        error: e.toString(),
      );
    }
  }

  /// Process receipt image
  Future<SmartCaptureResult> processReceipt(File imageFile) async {
    try {
      final response = await _apiClient.uploadFile(
        '/smart-capture/receipt',
        filePath: imageFile.path,
        fieldName: 'file',
      );
      return SmartCaptureResult.fromJson(response.data);
    } catch (e) {
      return SmartCaptureResult(
        success: false,
        mode: CaptureMode.receipt,
        extractedData: {},
        error: e.toString(),
      );
    }
  }

  /// Process image based on capture mode
  Future<SmartCaptureResult> processImage(
    File imageFile,
    CaptureMode mode, {
    bool autoCreate = false,
    String? linkedEntityId,
    String? linkedEntityType,
  }) async {
    switch (mode) {
      case CaptureMode.businessCard:
        return processBusinessCard(imageFile, autoCreate: autoCreate);
      case CaptureMode.document:
        return processDocument(imageFile);
      case CaptureMode.handwritten:
        return processHandwritten(
          imageFile,
          createNote: autoCreate,
          linkedEntityId: linkedEntityId,
          linkedEntityType: linkedEntityType,
        );
      case CaptureMode.receipt:
        return processReceipt(imageFile);
    }
  }

  /// Create CRM entity from extracted data
  Future<SmartCaptureResult> createEntityFromCapture({
    required CrmEntityType entityType,
    required Map<String, dynamic> extractedData,
    String? linkedEntityId,
    String? linkedEntityType,
    Map<String, dynamic>? additionalFields,
  }) async {
    try {
      final response = await _apiClient.post(
        '/smart-capture/create-entity',
        data: {
          'entityType': entityType.name,
          'extractedData': extractedData,
          'linkedEntityId': linkedEntityId,
          'linkedEntityType': linkedEntityType,
          'additionalFields': additionalFields,
        },
      );
      return SmartCaptureResult.fromJson(response.data);
    } catch (e) {
      return SmartCaptureResult(
        success: false,
        mode: CaptureMode.document,
        extractedData: extractedData,
        error: e.toString(),
      );
    }
  }
}

/// Provider for SmartCaptureService
final smartCaptureServiceProvider = Provider<SmartCaptureService>((ref) {
  final apiClient = ref.watch(apiClientProvider);
  return SmartCaptureService(apiClient);
});

/// Provider for OCR capabilities
final ocrCapabilitiesProvider = FutureProvider<OcrCapabilities>((ref) async {
  final service = ref.watch(smartCaptureServiceProvider);
  return service.getCapabilities();
});
